transports/janus_websockets.c

```diff
@@ -0,0 +1,906 @@
+	struct libwebsocket *wsi;				/* The libwebsock client instance */
+	GAsyncQueue *messages;					/* Queue of outgoing messages to push */
+	janus_mutex mutex;						/* Mutex to lock/unlock this session */
+	gint session_timeout:1;					/* Whether a Janus session timeout occurred in the core */
+	gint destroy:1;							/* Flag to trigger a lazy session destruction */
+} janus_websockets_client;
+
+
+/* libwebsockets WS context(s) */
+static struct libwebsocket_context *wss = NULL, *swss = NULL,
+	*admin_wss = NULL, *admin_swss = NULL;
+/* libwebsockets sessions that have been closed */
+static GList *old_wss;
+static janus_mutex old_wss_mutex;
+/* Callback for HTTP-related events (automatically rejected) */
+static int janus_websockets_callback_http(struct libwebsocket_context *this, 
+		struct libwebsocket *wsi,
+		enum libwebsocket_callback_reasons reason,
+		void *user, void *in, size_t len);
+/* Callbacks for WebSockets-related events */
+static int janus_websockets_callback(struct libwebsocket_context *this, 
+		struct libwebsocket *wsi,
+		enum libwebsocket_callback_reasons reason,
+		void *user, void *in, size_t len);
+static int janus_websockets_admin_callback(struct libwebsocket_context *this, 
+		struct libwebsocket *wsi,
+		enum libwebsocket_callback_reasons reason,
+		void *user, void *in, size_t len);
+/* Protocol mappings */
+static struct libwebsocket_protocols wss_protocols[] = {
+	{ "http-only", janus_websockets_callback_http, 0, 0 },
+	{ "janus-protocol", janus_websockets_callback, sizeof(janus_websockets_client), 0 },
+	{ NULL, NULL, 0 }
+};
+static struct libwebsocket_protocols admin_wss_protocols[] = {
+	{ "http-only", janus_websockets_callback_http, 0, 0 },
+	{ "janus-admin-protocol", janus_websockets_admin_callback, sizeof(janus_websockets_client), 0 },
+	{ NULL, NULL, 0 }
+};
+/* Helper for debugging reasons */
+const char *janus_websockets_reason_string(enum libwebsocket_callback_reasons reason);
+const char *janus_websockets_reason_string(enum libwebsocket_callback_reasons reason) {
+	switch(reason) {
+		case LWS_CALLBACK_ESTABLISHED:
+			return "LWS_CALLBACK_ESTABLISHED";
+		case LWS_CALLBACK_CLIENT_CONNECTION_ERROR:
+			return "LWS_CALLBACK_CLIENT_CONNECTION_ERROR";
+		case LWS_CALLBACK_CLIENT_FILTER_PRE_ESTABLISH:
+			return "LWS_CALLBACK_CLIENT_FILTER_PRE_ESTABLISH";
+		case LWS_CALLBACK_CLIENT_ESTABLISHED:
+			return "LWS_CALLBACK_CLIENT_ESTABLISHED";
+		case LWS_CALLBACK_CLOSED:
+			return "LWS_CALLBACK_CLOSED";
+		case LWS_CALLBACK_CLOSED_HTTP:
+			return "LWS_CALLBACK_CLOSED_HTTP";
+		case LWS_CALLBACK_RECEIVE:
+			return "LWS_CALLBACK_RECEIVE";
+		case LWS_CALLBACK_CLIENT_RECEIVE:
+			return "LWS_CALLBACK_CLIENT_RECEIVE";
+		case LWS_CALLBACK_CLIENT_RECEIVE_PONG:
+			return "LWS_CALLBACK_CLIENT_RECEIVE_PONG";
+		case LWS_CALLBACK_CLIENT_WRITEABLE:
+			return "LWS_CALLBACK_CLIENT_WRITEABLE";
+		case LWS_CALLBACK_SERVER_WRITEABLE:
+			return "LWS_CALLBACK_SERVER_WRITEABLE";
```

Comment:
I was honestly not aware of this approach, which I'm only now learning is called something like "stringize"... you never stop learning! Thanks, this will come in handy in other places too :smile: